### ISSUE TITLE
How to understand Null_Logits in Classifier Free Guidance?

### ISSUE THREAD

**Issue by Leo-T-Zang** [lucidrains/denoising-diffusion-pytorch#184]
Hi Phil,

In your classifier free guidance file, you use `self.null_classes_emb = nn.Parameter(torch.randn(dim))`, which is random control information.

In `forward_with_cond_scale` and `forward`
```
        null_logits = self.forward(*args, cond_drop_prob = 1., **kwargs)
        return null_logits + (logits - null_logits) * cond_scale
```
```
        if cond_drop_prob > 0:
            keep_mask = prob_mask_like((batch,), 1 - cond_drop_prob, device = device)
            null_classes_emb = repeat(self.null_classes_emb, 'd -> b d', b = batch)

            classes_emb = torch.where(
                rearrange(keep_mask, 'b -> b 1'),
                classes_emb,
                null_classes_emb
            )
```
My understanding is when `cond_drop_prob = 1`, these codes return x totally conditioned on null_classes_emb, which is random information. But why are we training model to predict x based on random information guidance? Instead just set `class_emb` zeros, which is unconditional?

Also, I would like to attach equation from paper [Classifier-Free Diffusion Guidance](https://arxiv.org/abs/2207.12598) for reference.
![image](https://user-images.githubusercontent.com/85547488/224546007-04f6965d-25f4-4952-b87d-14a745003efc.png)

Thank you.

**Comment by TreeberryTomato** [comment]
Hi, i have the same question, have you figured it out?
